### PR TITLE
Rework the setting and description of system clock to the epoch

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: man/systemd.xml [hunk 1]
```diff
@@ -1487,7 +1485,26 @@
   </refsect1>
 
   <refsect1>
-    <title>Sockets and FIFOs</title>
+    <title>System clock epoch</title>
+
+    <para>When <command>systemd</command> is started or restarted, it may set the system clock to the
+    "epoch". This mechanism is used to ensure that the system clock remains somewhat reasonably initialized
+    and roughly monotonic across reboots, in case no battery-backed local RTC is available or it does not
```
Comment (t184256) [inline man/systemd.xml:1492]: If the goal is to keep it monotonic *across reboots*, can first boots be excluded from this heuristics?

Scenario I have in mind: fresh installations, e.g., a software archeologist booting a CentOS 7 machine to compile a manylinux1 wheel about a decade from now.

File: src/shared/clock-util.c [hunk 2]
```diff
@@ -133,21 +133,27 @@ int clock_reset_timewarp(void) {
...
+                log_warning_errno(errno, "Could not stat %s, ignoring: %m", TIMESYNCD_CLOCK_FILE);
+
+        if (stat(EPOCH_CLOCK_FILE, &st) >= 0)
+                timesyncd_usec = timespec_load(&st.st_mtim);
+        else if (errno != ENOENT)
+                log_warning_errno(errno, "Could not stat %s, ignoring: %m", EPOCH_CLOCK_FILE);
+
+        epoch_usec = MAX3(epoch_usec,
+                          timesyncd_usec,
+                          (usec_t) TIME_EPOCH * USEC_PER_SEC);
```
Comment (poettering) [inline src/shared/clock-util.c]: if TIME_EPOCH is zero, and the files don't exist we'll actually reset the clock to zero even if valid. That's a bit weird.

Maybe just explicitly check for a zero epoch_usec below and turn it into CLOCK_CHANGE_NOOP

File: src/timesync/timesyncd.c [hunk 3]
```diff
@@ -22,9 +22,8 @@
 #include "timesyncd-manager.h"
 #include "user-util.h"
 
-static int advance_tstamp(int fd, const struct stat *st) {
+static int advance_tstamp(int fd, usec_t min) {
```
Comment (poettering) [inline src/timesync/timesyncd.c]: i'd probably not call this "min", since various libraries have macros of such a name

**Comment by poettering** [comment]
hmm, i can't say i like the "reaching over" of pid1 into timesyncd's touch file. If you want pid1 to honour that file too then maybe move it into a generic place, outside of timesyncd's private state dir.

/var/lib/systemd/clock-epoch?

File: man/systemd.xml (hunk 1 above)
Comment (keszybz) [reply man/systemd.xml:1492]: I don't think this would be a good idea. We certainly want to set the clock on first boot.

The archeologist will have to install an old version of systemd, reset the time explicitly after boot, or compile systemd with an old epoch set.

File: src/shared/clock-util.c (hunk 2 above)
Comment (keszybz) [reply src/shared/clock-util.c]: It's a bit of a corner case, but people occasionally do explicitly set mtimes to 0, so TIME_EPOCH could realistically be 0. I added an explicit check.

**Comment by keszybz** [comment]
> i can't say i like the "reaching over" of pid1 into timesyncd's touch file. If you want pid1 to honour that file too then maybe move it into a generic place, outside of timesyncd's private state dir.

If you insist, I'll update the patch to move the file, but I don't think we should do that. Both programs are compiled together and under our control, so there is really no functional difference between the paths. But if the file is moved, then we have to have logic for the transition, taking into account both locations, possibly cleaning up the old location, etc. And we create work for others. In particular, I expect that selinux and apparmor policies would need to be updated. So it's quite a bit of work for no change in functionality.

**Comment by keszybz** [comment]
Updated with the other requested changes, except as commented above.

**Comment by bluca** [comment]
> > i can't say i like the "reaching over" of pid1 into timesyncd's touch file. If you want pid1 to honour that file too then maybe move it into a generic place, outside of timesyncd's private state dir.
> 
> If you insist, I'll update the patch to move the file, but I don't think we should do that. Both programs are compiled together and under our control, so there is really no functional difference between the paths. But if the file is moved, then we have to have logic for the transition, taking into account both locations, possibly cleaning up the old location, etc. And we create work for others. In particular, I expect that selinux and apparmor policies would need to be updated. So it's quite a bit of work for no change in functionality.

Yeah I don't see an advantage to move this either, and several disadvantages and churn

**Comment by keszybz** [comment]
noble-s390x ran out of disk space and noble-arm64 failed in networkd tests.

File: man/systemd.xml [hunk 4]
```diff
@@ -1482,7 +1485,26 @@
...
+    work correctly.</para>
+
+    <para>The epoch is the lowest date below above which the system clock time is assumed to be set
```
Comment (bluca) [inline man/systemd.xml]: "lowest date above which"

**Review by bluca — APPROVED** [review]

File: man/systemd.xml (hunk 4 above)
Comment (keszybz) [reply man/systemd.xml]: Arrgh. I tried to reword this too many times. I'll push in a moment. I think we don't have to run CI again. It seems to be passing now.

Comment (keszybz) [reply man/systemd.xml]: Actually, this PR probably shouldn't be merged before the release. So the CI will have plenty of time.

Comment (keszybz) [reply man/systemd.xml]: … and it's even marked properly. Just ignore my rambling please.

File: src/core/main.c [hunk 5]
```diff
@@ -1704,24 +1704,33 @@ static void initialize_clock(void) {
...
         else if (r > 0 && change_dir == CLOCK_CHANGE_BACKWARD)
-                log_info("System time is further ahead than %s after build time, resetting clock to build time.",
+                log_info("System time is further ahead than %s after epoch, resetting clock to epoch.",
                          FORMAT_TIMESPAN(CLOCK_VALID_RANGE_USEC_MAX, USEC_PER_DAY));
         else if (r < 0 && change_dir == CLOCK_CHANGE_FORWARD)
-                log_error_errno(r, "Current system time is before build time, but cannot correct: %m");
+                log_error_errno(r, "Current system time is before epoch, but cannot correct: %m");
         else if (r < 0 && change_dir == CLOCK_CHANGE_BACKWARD)
-                log_error_errno(r, "Current system time is further ahead %s after build time, but cannot correct: %m",
+                log_error_errno(r, "Current system time is further ahead %s after epoch, but cannot correct: %m",
```
Comment (YHNdnzj) [inline src/core/main.c]: Hmm, if you touch this, maybe use `switch (change_dir)`? Also, please make the error msg and normal log msg consistent (`System time is further ahead _than_ %s` vs `system time is further ahead %s`

File: src/timesync/timesyncd.c [hunk 6]
```diff
@@ -22,9 +22,8 @@
 #include "timesyncd-manager.h"
 #include "user-util.h"
 
-static int advance_tstamp(int fd, const struct stat *st) {
+static int advance_tstamp(int fd, usec_t epoch) {
         assert_se(fd >= 0);
```
Comment (YHNdnzj) [inline src/timesync/timesyncd.c]: Not your fault, but assert_se -> assert

File: src/timesync/timesyncd.c [hunk 7]
```diff
@@ -91,49 +88,47 @@ static int load_clock_timestamp(uid_t uid, gid_t gid) {
...
                 return 0;
         }
 
         log_struct(LOG_INFO,
                    "MESSAGE_ID=" SD_MESSAGE_TIME_BUMP_STR,
-                   "REALTIME_USEC=" USEC_FMT, min+1,
-                   LOG_MESSAGE("System clock time unset or jumped backwards, restored from recorded timestamp: %s",
-                               FORMAT_TIMESTAMP(min+1)));
+                   "REALTIME_USEC=" USEC_FMT, epoch + 1,
+                   LOG_MESSAGE("System clock time advanced to %s: %s",
```
Comment (YHNdnzj) [inline src/timesync/timesyncd.c:130]: Please update the corresponding catalog entry. Plus, I think we should use the structured message in pid1 too.

File: man/systemd.xml [hunk 8]
```diff
@@ -1487,7 +1485,26 @@
...
+
+    <para>When <command>systemd</command> is started or restarted, it may set the system clock to the
+    "epoch". This mechanism is used to ensure that the system clock remains somewhat reasonably initialized
+    and roughly monotonic across reboots, in case no battery-backed local RTC is available or it does not
+    work correctly.</para>
+
+    <para>The epoch is the lowest date above which the system clock time is assumed to be set correctly. When
+    initializing, the local clock is <emphasis>advanced</emphasis> to the epoch if it was set to a lower
+    value. As a special case, if the local clock is sufficiently far in the future (by default 15 years, but
+    this can be configured at build time), the hardware clock is assumed to be broken, and the system clock
```
Comment (YHNdnzj) [inline man/systemd.xml:1498]: Let's just insert the build-time default here?

File: src/core/main.c (hunk 5 above)
Comment (keszybz) [reply src/core/main.c]: This is C, not Rust or Python! We can't have conditionalized matches, and without that a switch would be very awkward.

File: src/timesync/timesyncd.c (hunk 7 above)
Comment (keszybz) [reply src/timesync/timesyncd.c:130]: Actually, the catalog entry is already correct. timesyncd only advances the clock based on a timestamp. It's systemd that has more complicated logic.

File: src/timedate/hwclock-util.c [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+#include <fcntl.h>
+#include <linux/rtc.h>
+#include <sys/ioctl.h>
+#include <sys/time.h>
+
+#include "errno-util.h"
+#include "fd-util.h"
+#include "hwclock-util.h"
+
+int hwclock_get(struct tm *tm) {
```
Comment (poettering) [inline src/timedate/hwclock-util.c]: rename parameter to `ret`

File: src/core/clock-warp.c [hunk 10]
```diff
@@ -19,18 +19,28 @@ int clock_reset_timewarp(void) {
...
 void clock_apply_epoch(void) {
-        usec_t epoch_usec;
+        usec_t epoch_usec = 0, timesyncd_usec = 0;
         struct stat st;
         int r;
 
+        r = RET_NERRNO(stat(TIMESYNCD_CLOCK_FILE, &st));
+        if (r >= 0)
+                epoch_usec = timespec_load(&st.st_mtim);
+        else if (errno != ENOENT)
```
Comment (poettering) [inline src/core/clock-warp.c]: you want to check `r` here i guess

File: src/core/clock-warp.c [hunk 11]
```diff
@@ -19,18 +19,28 @@ int clock_reset_timewarp(void) {
...
+                log_warning_errno(r, "Could not stat %s, ignoring: %m", TIMESYNCD_CLOCK_FILE);
+
         r = RET_NERRNO(stat(EPOCH_CLOCK_FILE, &st));
-        if (r < 0) {
-                if (r != -ENOENT)
-                        log_warning_errno(r, "Cannot stat " EPOCH_CLOCK_FILE ": %m");
+        if (r >= 0)
+                timesyncd_usec = timespec_load(&st.st_mtim);
+        else if (errno != ENOENT)
```
Comment (poettering) [inline src/core/clock-warp.c]: ditto

File: src/core/clock-warp.c [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        if (r < 0 && advance)
+                return (void) log_error_errno(r, "Current system time is before build time, but cannot correct: %m");
+        else if (r < 0)
+                return (void) log_error_errno(r, "Current system time is further ahead than %s after build time, but cannot correct: %m",
+                                              FORMAT_TIMESPAN(CLOCK_VALID_RANGE_USEC_MAX, USEC_PER_DAY));
+        else if (advance)
+                log_info("System time was before build time, advanced clock.");
+        else
+                log_info("System time was further ahead than %s after build time, reset clock to build time.",
+                         FORMAT_TIMESPAN(CLOCK_VALID_RANGE_USEC_MAX, USEC_PER_DAY));
```
Comment (poettering) [inline src/core/clock-warp.c]: hmm you are doing one test twice here. why not just:

```c

if (r < 0) {
        if (advance)
                log_error_errno(r, "…
        else 
                log_error_errno(r, "…
        return;
}
if (advance) 
        …
else 
        …

File: man/systemd.xml (hunk 8 above)
Comment (keszybz) [reply man/systemd.xml:1498]: The problem with that is that the value is in µs, and it's a huge number. To show it to the user, we'd need to format it somehow, and we don't have the infra for that. I don't think think this is important enough to add special handling for this.

File: src/core/clock-warp.c (hunk 12 above)
Comment (keszybz) [reply src/core/clock-warp.c]: The compiler can surely optimize this away. I don't want to have an additional level of indentation.

**Comment by keszybz** [comment]
Updated.

File: src/core/clock-warp.c (hunk 12 above)
Comment (YHNdnzj) [reply src/core/clock-warp.c]: Well, I think adding one indentation here makes things more readable. So I agree with @poettering.

Comment (poettering) [reply src/core/clock-warp.c]: dunno. i find it more readable if we clearly separate error handling from success handling

File: src/systemd/sd-messages.h [hunk 13]
```diff
@@ -252,6 +252,9 @@ _SD_BEGIN_DECLARATIONS;
 #define SD_MESSAGE_TIME_BUMP                          SD_ID128_MAKE(7d,b7,3c,8a,f0,d9,4e,eb,82,2a,e0,43,23,fe,6a,b6)
 #define SD_MESSAGE_TIME_BUMP_STR                      SD_ID128_MAKE_STR(7d,b7,3c,8a,f0,d9,4e,eb,82,2a,e0,43,23,fe,6a,b6)
 
+#define SD_MESSAGE_TIME_SET_EPOCH                     SD_ID128_MAKE(d9,3d,93,46,71,97,42,bd,ad,92,75,df,52,66,cb,58)
+#define SD_MESSAGE_TIME_SET_EPOCH_STR                 SD_ID128_MAKE_STR(d9,3d,93,46,71,97,42,bd,ad,92,75,df,52,66,cb,58)
```
Comment (YHNdnzj) [inline src/systemd/sd-messages.h]: Hmm, this has no associated catalog entry?

What I meant is to reuse the time bump log id (with some appropriate rewording of cource)

File: src/shared/clock-util.c [hunk 14]
```diff
@@ -69,9 +29,8 @@ int clock_is_localtime(const char* adjtime_path) {
         f = fopen(adjtime_path, "re");
         if (f) {
```
Comment (YHNdnzj) [inline src/shared/clock-util.c]: If you touch this, could you please do error check here first?

File: src/shared/clock-util.c [hunk 15]
```diff
@@ -18,40 +12,6 @@
...
-        assert(tm);
-
-        fd = open("/dev/rtc", O_RDONLY|O_CLOEXEC);
-        if (fd < 0)
-                return -errno;
-
-        return RET_NERRNO(ioctl(fd, RTC_SET_TIME, tm));
-}
-
 int clock_is_localtime(const char* adjtime_path) {
```
Comment (YHNdnzj) [inline src/shared/clock-util.c]:
```suggestion
int clock_is_localtime(const char *adjtime_path) {
```

**Review by YHNdnzj — APPROVED** [review]
Overall looks pretty good, but still needs one more round.

File: src/core/clock-warp.c (hunk 12 above)
Comment (keszybz) [reply src/core/clock-warp.c]: OK, I changed it like that.

File: src/systemd/sd-messages.h (hunk 13 above)
Comment (keszybz) [reply src/systemd/sd-messages.h]: OK, updated.

Comment (keszybz) [reply src/systemd/sd-messages.h]: I dropped the catalog entry because it's fully useless now.
